metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java

```diff
@@ -230,138 +224,221 @@ private static void setDereferenceStatusInReport(String resourceId, HashSet<Repo
+    resourceIds.forEach((key, value) -> {
+      HashSet<Report> reports = new HashSet<>();
+      Set<ReferenceTerm> referenceTermSet = setUpReferenceTermSet(value, reports);
+      mappedReferenceTerms.put(key, referenceTermSet);
+      final DereferencedEntities dereferencedOwnEntities = dereferenceOwnEntities(referenceTermSet, reports, key);
+      dereferencedResultEntities.put(key, dereferencedOwnEntities);
+    });
+
+    final Set<String> foundOwnEntityIds = dereferencedResultEntities
+        .values().stream()
+        .map(DereferencedEntities::getReferenceTermListMap)
+        .map(Map::values).flatMap(Collection::stream).flatMap(Collection::stream)
+        .map(EnrichmentBase::getAbout)
         .collect(Collectors.toSet());
 
-    final DereferencedEntity deferencedOwnEntities = dereferenceOwnEntities(referenceTermSet);
-    reports.addAll(deferencedOwnEntities.getReportMessages());
-
-    final Set<String> foundOwnEntityIds = deferencedOwnEntities.getEnrichmentBaseList()
-                                                               .stream()
-                                                               .map(EnrichmentBase::getAbout)
-                                                               .collect(Collectors.toSet());
-
     // For the remaining ones, get them from the dereference service.
-    for (ReferenceTerm resourceId : referenceTermSet) {
-      if (!foundOwnEntityIds.contains(resourceId.getReference().toString())) {
-        DereferencedEntity deferencedExternalEntities =
-            dereferenceExternalEntity(resourceId.getReference().toString());
-        reports.addAll(deferencedExternalEntities.getReportMessages());
-        deferencedOwnEntities.getEnrichmentBaseList().addAll(deferencedExternalEntities.getEnrichmentBaseList());
+    for (Map.Entry<Class<? extends AboutType>, Set<ReferenceTerm>> entry : mappedReferenceTerms.entrySet()) {
+      DereferencedEntities dereferencedEntities;
+      Set<ReferenceTerm> notFoundOwnReferenceTerms = entry.getValue().stream().filter(
+          referenceTerm -> !foundOwnEntityIds.contains(referenceTerm.getReference().toString())).collect(
+          Collectors.toSet());
+      if(notFoundOwnReferenceTerms.isEmpty()){
+        continue;
       }
+
+      if (entry.getKey().equals(Aggregation.class)) {
+        dereferencedEntities = dereferenceAggregation(notFoundOwnReferenceTerms, entry.getKey());
+      } else {
+        dereferencedEntities = dereferenceExternalEntity(notFoundOwnReferenceTerms, entry.getKey());
+      }
+
+      updateDereferencedEntitiesMap(dereferencedResultEntities, entry.getKey(), dereferencedEntities);
     }
     // Done.
-    return new DereferencedEntity(deferencedOwnEntities.getEnrichmentBaseList(), reports);
+    return new ArrayList<>(dereferencedResultEntities.values());
+  }
+
+  private DereferencedEntities dereferenceAggregation(Set<ReferenceTerm> referenceTerms, Class<? extends AboutType> classType) {
```

Comment:
This seems way better now but I think I'm seeing something that would misbehave, please bare with me :D  
So let's say we have a Set(nonEuropeanaURI1, nonEuropeanaURI1).  
The result could then contain two values of Europeana entities for the first nonEuropeanaURI1.  
So we'd have a result(EuropeanaURI1, EuropeanaURI2).  
Then we create the remainingReferences which would be the references in the input set that are not in the result, which will be both nonEuropeanaURI1 and nonEuropeanaURI1(since they are both not in the result).  
Now they would be both dereferenced from an external provider, pontentially adding duplicate values in the result(some from the Entity API and some from External)?
Is that true?

Reply:
Accord to the javadoc for `resolveByUri` in class `EntityResolver`, this method returns `A map from provided references to the entities that the references are equivalent to.` So if not mistaken, the keys in the map would be the reference terms we provided. 
So the result would be <nonEuropeanaURI1, (EuropeanaURI1, EuropeanaURI2)>. and since for the `remainingReferences` we check if the key set of `result` contains any of provided `referenceTerms`, in the end the `remainingReferences` will be (nonEuropeanaURI2) only. 
At least this is my understanding, I can also test this later and make sure is working

Reply:
You absolutely right, I missed one of the instructions which lead me to wrong logic.
That being said though, I had a look further and indeed in the `EntityResolver` says `Note: not all provided references may be present (i.e. if the reference didn't resolve).` which then all makes sense but I checked the `ClientEntityResolver` and it seems that in the result there would also be keys in the map(with empty list as value) for those that no entity was found:
https://github.com/europeana/metis-framework/blob/47307cd76c46b2857a34a1ac6a218ed45ce1106f/metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/impl/ClientEntityResolver.java#L106

Can we verify that?